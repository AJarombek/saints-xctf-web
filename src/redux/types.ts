/**
 * Types for the Redux store.
 * @author Andrew Jarombek
 * @since 7/23/2020
 */

export type RootState = {
    auth: AuthState,
    logs: LogsState
}

export type LogsState = {
    isFetching: boolean;
    didInvalidate: boolean;
    lastUpdated: number;
    items: Logs;
    feeds: LogFeeds;
    newComments: NewComments;
}

export type AuthState = {
    auth: Auth;
    user: Users;
}

export interface Meta {
    isFetching?: boolean;
    lastUpdated?: number;
    didInvalidate?: boolean;
    serverError?: string;
}

export type Auth = {
    isFetching?: boolean;
    lastUpdated?: number;
    signedInUser?: string;
    status?: string;
}

export type Users = {
    [key: string]: UserMeta;
}

export interface UserMeta extends User, Meta {}

export interface User {
    activation_code?: string;
    class_year?: number;
    deleted?: string;
    description?: string;
    email?: string;
    favorite_event?: string;
    first?: string;
    last?: string;
    last_signin?: string;
    location?: string;
    member_since?: string;
    password?: string;
    salt?: string;
    subscribed?: string;
    username?: string;
    week_start?: string;
}

enum LogType { RUN = "run", BIKE = "bike", SWIM = "swim", OTHER = "other" }
enum Metric { MILES = "miles", KILOMETERS = "kilometers", METERS = "meters" }

export type Log = {
    log_id: number;
    username: string;
    first: string;
    last: string;
    name: string;
    date: string;
    type: LogType;
    distance?: number;
    metric?: Metric;
    miles?: number;
    time?: string;
    pace?: string;
    feel: number;
    location?: string;
    description?: string;
    timeCreated?: string;
    comments?: Comment[];
}

export type Logs = {
    [key: string]: Record<string, Log>;
}

export type LogFeeds = {
    [key: string]: {
        filterBy: string;
        bucket: string;
        pages: LogFeedPages
    };
}

export type LogFeedPages = {
    [key: string]: LogFeedPage
}

export type LogFeedPage = {
    isFetching: boolean;
    lastUpdated: number;
    items: Log[];
    serverError: string;
}

export type Comment = {
    comment_id: number;
    username: string;
    first: string;
    last: string;
    log_id: number;
    time: string;
    content: string;
}

export type NewComments = {
    [key: string]: {
        isFetching: boolean;
        lastUpdated: number;
        created?: boolean;
        serverError?: string;
    }
}
