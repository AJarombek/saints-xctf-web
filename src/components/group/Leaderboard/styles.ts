/**
 * JSS styles for the Leaderboard component.
 * @author Andrew Jarombek
 * @since 1/10/2021
 */

import { Modules } from '../../../styles/modules';
import { AJComponentMixins } from '../../../styles/mixins';
import Colors from '../../../styles/colors';
import color from 'color';

export default {
  leaderboard: {},
  ...Modules.filters(),
  interval: {
    marginTop: 40,
    width: 120
  },
  select: {
    ...AJComponentMixins.ajSelect()
  },
  barChart: {
    backgroundColor: color(Colors.lightBackground).darken(0.05).hex(),
    padding: 20,
    borderRadius: 6,
    marginTop: 10
  },
  alertMessage: {
    marginTop: 10,

    '& > div': {
      padding: '4px 0'
    }
  },
  '@media screen and (max-width: 390px)': {
    barChart: {
      padding: 10
    },
    interval: {
      marginLeft: 15
    }
  }
};
