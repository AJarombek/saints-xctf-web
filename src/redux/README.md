### Overview

Configure Redux as the state management system for the application.

### Files

| Filename                 | Description                                                                |
|--------------------------|----------------------------------------------------------------------------|
| `modules`                | Redux modules containing Reducers, Actions, and Action Creators.           |
| `appState.js`            | Sample application state maintained by Redux.                              |
| `store.js`               | Configuration and creation of the Redux store.                             |
| `types.ts`               | Typescript types used throughout the application.                          |

### Resources

1) [React/Redux Example App #1](https://github.com/erikras/react-redux-universal-hot-example/tree/master/src/redux)
2) [React/Redux Example App #2](https://github.com/goopscoop/ga-react-tutorial/tree/6-reduxActionsAndReducers/src/redux)
3) [React/Redux Example App #3](https://github.com/gothinkster/react-redux-realworld-example-app/tree/master/src)
4) [Redux Ducks Spec](https://github.com/erikras/ducks-modular-redux)
5) [React Router/Redux Setup Steps](https://github.com/supasate/connected-react-router#usage)
6) [Redux Hot Reloading](https://redux.js.org/recipes/configuring-your-store/#hot-reloading)
7) [Redux Thunk](https://github.com/reduxjs/redux-thunk)
8) [Redux/React Official Real World Example](https://github.com/reduxjs/redux/tree/master/examples/real-world)
