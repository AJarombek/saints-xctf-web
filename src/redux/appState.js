/**
 * Sample Redux State for the entire application.
 * @author Andrew Jarombek
 * @since 5/3/2020
 */

const appState = {
  auth: {
    auth: {
      isFetching: false,
      lastUpdated: 1588530309,
      signedIn: true,
      status: "SUCCESS"
    },
    registration: {
      stage: 1,
      first: "Andy",
      last: "Jarombek",
      email: "andrew@jarombek.com",
      username: undefined,
      password: undefined,
      activationCode: undefined,
      teams: {
        saintsXCTFAlumni: {
          status: "accepted",
          groups: {
            alumni: {
              status: "accepted"
            }
          }
        }
      }
    },
    user: {
      andy: {
        isFetching: false,
        didInvalidate: false,
        lastUpdated: 1588530310,
        username: "andy",
        first: "Andrew",
        last: "Jarombek",
        password: "xxxxxx",
        description: "Test Description",
        memberSince: "2016-02-23 12:00:00",
        classYear: 2017,
        location: "New York, NY",
        favoriteEvent: "Cooldown Jog",
        activationCode: "ABCD1234",
        email: "andrew@jarombek.com",
        lastSignIn: "2020-05-03 12:00:00",
        subscribed: "Y",
        logs: []
      }
    },
  },
  profile: {
    tom: {
      isFetching: false,
      didInvalidate: false,
      lastUpdated: 1588530311
    },
    joe: {
      isFetching: false,
      didInvalidate: false,
      lastUpdated: 1588530312
    },
    ben: {
      isFetching: false,
      didInvalidate: false,
      lastUpdated: 1588530313
    }
  },
  group: {
    wmenstf: {
      isFetching: false,
      didInvalidate: false,
      lastUpdated: 1588530314,
      groupName: "wmenstf",
      groupTitle: "Women's Track & Field",
      description: "",
      weekStart: "monday",
      members: [],
      logs: []
    }
  },
  logs: {
    isFetching: false,
    didInvalidate: false,
    lastUpdated: 1588530315,
    items: {
      1: {
        log_id: 1,
        username: "andy",
        first: "Andrew",
        last: "Jarombek",
        name: "Reservoir Loop",
        location: "New York, NY",
        date: "2020-05-03",
        type: "run",
        distance: 6.37,
        metric: "miles",
        miles: 6.37,
        time: "00:45:37",
        pace: "00:07:10",
        feel: 6,
        description: "",
        timeCreated: "2020-05-03 12:00:00",
        comments: {
          1: {
            commentId: 1,
            username: "andy",
            first: "Andrew",
            last: "Jarombek",
            logId: 1,
            time: "2020-05-03 12:00:00",
            content: "test"
          }
        }
      }
    }
  }
};
