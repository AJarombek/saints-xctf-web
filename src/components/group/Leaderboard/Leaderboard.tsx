/**
 * Component for a tab on the group page which shows a leaderboard of the group members exercises.
 * @author Andrew Jarombek
 * @since 1/10/2021
 */

import React, { useEffect, useMemo, useState } from 'react';
import { createUseStyles } from 'react-jss';
import styles from './styles';
import { useDispatch, useSelector } from 'react-redux';
import {
  ExerciseFilters,
  GroupMeta,
  LeaderboardInterval,
  LeaderboardItem as LeaderboardItemType,
  LeaderboardItemMeta,
  Leaderboards,
  RootState,
} from '../../../redux/types';
import { getGroupLeaderboard } from '../../../redux/modules/groups';
import FilterButtons from '../../shared/FilterButtons';
import { AJSelect } from 'jarombek-react-components';
import LeaderboardItem from '../LeaderboardItem';
import Alert from '../../shared/Alert';
import classNames from 'classnames';

const useStyles = createUseStyles(styles);

export type CurrentLeaderboardItem = {
  username: string;
  first: string;
  last: string;
  value: number;
};

interface Props {
  group: GroupMeta;
}

const Leaderboard: React.FunctionComponent<Props> = ({ group }) => {
  const classes = useStyles();

  const dispatch = useDispatch();
  const allLeaderboards: Record<string, Leaderboards> = useSelector((state: RootState) => state.groups.leaderboards);

  const [interval, setInterval] = useState<LeaderboardInterval>('week');

  const [selectedFilters, setSelectedFilters] = useState<ExerciseFilters>({
    run: true,
    bike: false,
    swim: false,
    other: false,
  });

  useEffect(() => {
    if (group?.id && !(allLeaderboards[group.id] && allLeaderboards[group.id][interval])) {
      dispatch(getGroupLeaderboard(group.id, interval));
    }
  }, [group, dispatch, interval, allLeaderboards]);

  const leaderboardItemMeta: LeaderboardItemMeta = useMemo(() => {
    if (group?.id) {
      const groupLeaderboard: Leaderboards = allLeaderboards[group.id] ?? {};
      return groupLeaderboard[interval];
    } else {
      return null;
    }
  }, [allLeaderboards, group.id, interval]);

  const leaderboardItems: LeaderboardItemType[] = useMemo(() => {
    return leaderboardItemMeta ? leaderboardItemMeta?.items : [];
  }, [leaderboardItemMeta]);

  const currentLeaderboard: CurrentLeaderboardItem[] = useMemo(() => {
    return (
      leaderboardItems
        ?.map((item: LeaderboardItemType) => ({
          username: item.username,
          first: item.first,
          last: item.last,
          value:
            (selectedFilters.run ? item.miles_run : 0) +
            (selectedFilters.bike ? item.miles_biked : 0) +
            (selectedFilters.swim ? item.miles_swam : 0) +
            (selectedFilters.other ? item.miles_other : 0),
        }))
        .sort((a: CurrentLeaderboardItem, b: CurrentLeaderboardItem) => b.value - a.value)
        .filter((item: CurrentLeaderboardItem) => item.value > 0) ?? []
    );
  }, [selectedFilters, leaderboardItems]);

  const leaderMiles: number = useMemo(() => {
    return currentLeaderboard.length ? currentLeaderboard[0].value : 0;
  }, [currentLeaderboard]);

  return (
    <div id="leaderboard" className={classes.leaderboard}>
      <div className={classes.filters}>
        <p className={classes.filterTitle}>Leaderboard Filters:</p>
        <FilterButtons selectedFilters={selectedFilters} setSelectedFilters={setSelectedFilters} />
      </div>
      <div className={classes.interval}>
        <AJSelect
          options={[
            { content: 'All Time', value: 'all' },
            { content: 'Past Year', value: 'year' },
            { content: 'Past Month', value: 'month' },
            { content: 'Past Week', value: 'week' },
          ]}
          defaultOption={4}
          onClickListOption={(item: { content: string; value: string }): void =>
            setInterval(item.value as LeaderboardInterval)
          }
          className={classNames(classes.select, 'leaderboardInterval')}
        />
      </div>
      {!!currentLeaderboard?.length && (
        <div className={classes.barChart}>
          {currentLeaderboard.map((item: CurrentLeaderboardItem) => (
            <LeaderboardItem key={item.username} item={item} leaderMiles={leaderMiles} />
          ))}
        </div>
      )}
      {!currentLeaderboard?.length &&
        leaderboardItemMeta &&
        !leaderboardItemMeta?.isFetching &&
        !!leaderboardItemMeta?.serverError && (
          <div className={classes.alertMessage} data-cypress="leaderboardAlert">
            <Alert message={leaderboardItemMeta.serverError} type="error" closeable={false} />
          </div>
        )}
      {!currentLeaderboard?.length &&
        leaderboardItemMeta &&
        !leaderboardItemMeta?.isFetching &&
        !!leaderboardItemMeta?.serverWarning && (
          <div className={classes.alertMessage} data-cypress="leaderboardAlert">
            <Alert message={leaderboardItemMeta.serverWarning} type="warning" closeable={false} />
          </div>
        )}
      {!currentLeaderboard?.length &&
        leaderboardItemMeta &&
        !leaderboardItemMeta?.isFetching &&
        !leaderboardItemMeta?.serverWarning &&
        !leaderboardItemMeta?.serverError && (
          <div className={classes.alertMessage} data-cypress="leaderboardAlert">
            <Alert
              message="There is no leaderboard data in this time interval with the current filters."
              type="warning"
              closeable={false}
            />
          </div>
        )}
    </div>
  );
};

export default Leaderboard;
