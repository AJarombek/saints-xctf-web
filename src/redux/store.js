/**
 * Build and configure the redux store, which holds the application state and handles
 * all state updates.
 * @author Andrew Jarombek
 * @since 5/1/2020
 */

import { createBrowserHistory } from 'history';
import { createStore, applyMiddleware } from 'redux';
import { routerMiddleware } from 'connected-react-router';
import createLogger from 'redux-logger';
import reducer from './modules/reducers';

const history = createBrowserHistory();

const configureStore = (data) => {
  const loggerMiddleware = createLogger();
  const middleware = [loggerMiddleware, routerMiddleware(history)];

  const createStoreWithMiddleware = applyMiddleware(...middleware)(createStore);
  return createStoreWithMiddleware(reducer, data);
};

export { history };
export default configureStore;
