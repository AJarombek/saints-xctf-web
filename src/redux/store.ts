/**
 * Build and configure the redux store, which holds the application state and handles
 * all state updates.
 * @author Andrew Jarombek
 * @since 5/1/2020
 */

import { createBrowserHistory } from 'history';
import { createStore, applyMiddleware, Store, Action } from 'redux';
import { routerMiddleware } from 'connected-react-router';
import { createLogger } from 'redux-logger';
import thunkMiddleware, { ThunkAction } from 'redux-thunk';
import reducer from './modules/reducers';
import { RootState } from './types';

export const history = createBrowserHistory();

export default function configureStore(): Store {
  const loggerMiddleware = createLogger();
  const middleware = [routerMiddleware(history), loggerMiddleware, thunkMiddleware];

  const createStoreWithMiddleware = applyMiddleware(...middleware)(createStore);
  const store = createStoreWithMiddleware(reducer(history));

  if (process.env.NODE_ENV !== 'production' && module.hot) {
    module.hot.accept('./modules/reducers', () => store.replaceReducer(reducer));
  }

  return store;
}

export type AppThunk<ReturnType = void, State = RootState> = ThunkAction<ReturnType, State, unknown, Action<string>>;
