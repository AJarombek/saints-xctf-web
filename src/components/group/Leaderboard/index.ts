import Leaderboard, { CurrentLeaderboardItem } from './Leaderboard';

export { CurrentLeaderboardItem };
export default Leaderboard;
