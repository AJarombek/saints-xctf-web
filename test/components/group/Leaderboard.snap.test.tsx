/**
 * Snapshot test for the Leaderboard component.
 * @author Andrew Jarombek
 * @since 3/4/2021
 */

import React from 'react';
import renderer from 'react-test-renderer';
import { Store } from 'redux';
import { emptyStore } from '../../test-utils/storeMocks';
import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { alumni } from '../../test-utils/groupMocks';
import Leaderboard from '../../../src/components/group/Leaderboard';

const mockStore = configureStore([thunk]);

describe('Leaderboard Snapshot Tests', () => {
  let store: Store;

  beforeEach(() => {
    store = mockStore(emptyStore);
  });

  it('renders correctly', () => {
    const tree = renderer
      .create(
        <MemoryRouter>
          <Provider store={store}>
            <Leaderboard group={alumni} />
          </Provider>
        </MemoryRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
