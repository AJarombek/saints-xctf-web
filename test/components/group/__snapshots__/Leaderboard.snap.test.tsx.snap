// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Leaderboard Snapshot Tests renders correctly 1`] = `
<div
  className="leaderboard-0-1-1"
>
  <div
    className="filters-0-1-2"
  >
    <p
      className="filterTitle-0-1-3"
    >
      Leaderboard Filters:
    </p>
    <div
      className="buttonGroup-2-2-1 filterButtons-0-1-8"
    >
      <div
        className="ajBaseButton-2-2-5 ajBaseButtonEnabled-2-2-6 aj-contained-button ajContainedButton-2-2-2 ajContainedButtonEnabled-2-2-3"
        onClick={[Function]}
      >
        <button
          disabled={false}
          type="button"
        >
          Run
        </button>
      </div>
      <div
        className="ajBaseButton-2-2-5 ajBaseButtonEnabled-2-2-6 aj-outlined-button ajOutlinedButton-2-2-8 ajOutlinedButtonEnabled-2-2-9"
        onClick={[Function]}
      >
        <button
          disabled={false}
          type="button"
        >
          Bike
        </button>
      </div>
      <div
        className="ajBaseButton-2-2-5 ajBaseButtonEnabled-2-2-6 aj-outlined-button ajOutlinedButton-2-2-8 ajOutlinedButtonEnabled-2-2-9"
        onClick={[Function]}
      >
        <button
          disabled={false}
          type="button"
        >
          Swim
        </button>
      </div>
      <div
        className="ajBaseButton-2-2-5 ajBaseButtonEnabled-2-2-6 aj-outlined-button ajOutlinedButton-2-2-8 ajOutlinedButtonEnabled-2-2-9"
        onClick={[Function]}
      >
        <button
          disabled={false}
          type="button"
        >
          Other
        </button>
      </div>
    </div>
  </div>
  <div
    className="interval-0-1-4"
  >
    <div
      className="aj-select ajSelect-2-2-11 aj-select-closed select-0-1-5"
    >
      <div
        className="ajSelectLabel-2-2-13"
        onClick={[Function]}
      >
        <div />
        <div
          className="ajSelectArrow-2-2-12"
        >
          C
        </div>
      </div>
    </div>
  </div>
</div>
`;
